OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+  /// function is called on an empty DataTable_, nrow must be provided.
+  /// Otherwise, nrow is ignored. To add just one col, use addRow().
+  ///
+  /// \param first Beginning of range covered by the iterator.
+  /// \param last End of the range covered by the iterator.
+  /// \param nrow Number of rows to create in the DataTable_. This is only
+  ///             used(and required) when the function is called on an empty 
+  ///             DataTable_. Ignore otherwise.
+  /// \param allow_missing Allow for missing values. When set to false, this
+  ///                      function will throw if the input iterator fills up 
+  ///                      the last col only partially. When set to true, 
+  ///                      missing elements are set to SimTK::NaN.
+  ///
+  /// \throws OpenSim::ZeroElements If the number of elements produced by the
+  ///                               input iterator is zero.
+  /// \throws OpenSim::InvalidEntry If the function is called on an empty
+  ///                               DataTable_ without providing the argument
+  ///                               nrow.
+  /// \throws OpenSim::InvalidEntry If allow_missing is false and the input
+  ///                               iterator does not fill up the last col
+  ///                               completely.
+  template<typename InputIt>
+  void addCols(InputIt first, 
+               InputIt last, 
+               const size_t nrow = std::numeric_limits<size_t>::max(),
+               const bool allow_missing = false) {
+    if(!(first != last))
+      throw ZeroElements{"Input iterators produce zero elements."};
+    if((data.nrow() == 0 || data.ncol() == 0) && 
+       (nrow == std::numeric_limits<size_t>::max() || nrow == 0))
+      throw InvalidEntry{"DataTable is empty. 'nrow' argument must be" 
+                         " provided and it cannot be zero."};
+
+    data.resizeKeep(data.nrow() == 0 ? static_cast<int>(nrow) : data.nrow(), 
+                    data.ncol() + 1);
+    int row{0};
+    int col{data.ncol() - 1};
+    while(first != last) {
+      data.set(row, col, *first);
+      ++first; ++row;
+      if(row == static_cast<int>(data.nrow()) && first != last) {
+        row = 0;
+        ++col;
+        data.resizeKeep(data.nrow(), data.ncol() + 1);
+      }
+    }
+    if(!allow_missing && row != data.nrow())
+      throw InvalidEntry{"Input iterator did not produce enough elements to "
+                         "fill the last col."};
+  }
+
+  /// Bind another DataTable_ to this DataTable_ by row. The new elements will 
+  /// appear as the last rows of this DataTable_. Only data will be appended.
+  /// Metadata is not added. Columns retain their labels. To create a new 
+  /// DataTable_  that is a bind of two existing DataTable_(s), see 
+  /// rbindDataTables().
+  ///
+  /// \throws OpenSim::InvalidEntry If input DataTable_ has incorrect number of
```

Comment:
I added documentation about exceptions because it tells user about valid inputs to a given function and is thus part of documentation. User can also be assured that we are doing error-checking for these situations. I can add to the class comment that these exceptions are not exhaustive.

My intention was not to just provide a list of exceptions thrown but also provide details on when it is thrown. This way, when a user gets an exception, he can refer back to the function documentation to learn a bit more. This is why I didn't provide nothrow in signatures.